### PR TITLE
[ENH + MRG] added symmetric_range kw to plot_stat_map

### PR DESCRIPTION
**Pull request by eickenberg**
Adding functionality to make colorbar range from 0 to vmax instead of -vmax to +vmax. Which example should I best illustrate this on? @lesteve

### CODE REVIEW

**Comment by bthirion** [comment]
The example looks good. I don't think you need to add more yticks. 
Could you add a (smoke) test ?

**Comment by eickenberg** [comment]
Added back the typical number of ticks for such a colorbar.

figure_1

Will add smoke test.

**Comment by eickenberg** [comment]
Travis failure due to smoke tests: Smoke tests for plots invoke saving in order to force rendering. Saving is forced by passing a filename. `img_plotting._plot_img_with_bg` saves the file if a filename is provided and returns `None` instead of the active display object. Changing the colorbar a posteriori in `plot_stat_map` by interacting with the display object is thus impossible and raises an error.

Two possibilities for solving this:

- add `cbar_min`, `cbar_max` kwargs to `_plot_img_with_bg`, making colorbar range editing available to a wider number of functions (is this wanted at all?)
- do the saving by hand in `plot_stat_map` after the colorbar has been rendered.

Neither of the possibilities poses much of a problem. It is more a question of whether specifying `cbar_max` and `cbar_min` independently of `vmax, vmin` is generally useful or not.

**Comment by AlexandreAbraham** [comment]
Option #3: turn _plot_img_with_bg into an object which plot the axis in the __init__, let you mess around with the colorbar, and render it (or save it to a file) in a .show() method. Note that I have not seen the code and therefore don't know if this make sense or not :confused: .

**Comment by lesteve** [comment]
> add cbar_min, cbar_max kwargs to _plot_img_with_bg, making colorbar range editing available to a wider number of functions (is this wanted at all?)

I'd be in favor of this solution even if the colorbar tweaking is only used in plot_stat_map. In my mind _plot_img_with_bg is the generic helper function that user-facing functions call with different parameters.

An other option would be to add a callback that is called with the display as argument just before the save. Not sure which one is simpler.

**Comment by GaelVaroquaux** [comment]
>   • add cbar_min, cbar_max kwargs to _plot_img_with_bg, making colorbar range
>     editing available to a wider number of functions (is this wanted at all?)

This sounds about right to me. _plot_img_with_bg is somewhat of a
low-level swiss-army knife, and can thus be made more versatile.

**Comment by eickenberg** [comment]
swiss army knife made and plot_stat_map tested

File: nilearn/plotting/img_plotting.py [hunk 1]
```diff
@@ -586,7 +611,15 @@ def plot_stat_map(stat_map_img, bg_img=MNI152TEMPLATE, cut_coords=None,
                     np.logical_not(stat_map_data._mask)])
         stat_map_max = np.nanmax(stat_map_data)
         stat_map_min = np.nanmin(stat_map_data)
+
+    if symmetric_cbar == 'auto':
+        symmetric_cbar = stat_map_min < 0
```
Comment (eickenberg) [inline nilearn/plotting/img_plotting.py]: i would, too. but is this expected to happen for any type of data?

Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: Yes. A t map with the contrast specified with the wrong sign for instance.

Comment (eickenberg) [inline nilearn/plotting/img_plotting.py]: but t values generally can take both signs and thus should be shown with
the full colorbar, but ok, i see the idea, and will add.

up until now i was thinking that showing only one half of the colorbar
should only be done if the values are necessarily and inherently of only
one sign (eg neg log p values). and in that case one has the choice of the
sign and won't let it be negative.

Comment (bthirion) [inline nilearn/plotting/img_plotting.py]: To be frank I don't think I have encountered such a use case in the past: while positive values van be seen for variance images, I have no idea of the equivalent for negative images.
On the other hand, I thinks that adding this use case could be useful in some cases that I can't figure out at the moment, and will not harm.

Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: Ok

**Comment by eickenberg** [comment]
Added possibility for fully negative range to be detected. Also made tick marks adapt to the necessary order of magnitude in precision (done using `log10` of the amplitude. Not switching from `%f` to `%e` notation and not invoking any general matplotlib tick formatter -- open to suggestions if such a thing exists).

**Comment by eickenberg** [comment]
Possibly found a recipe for the latter in matplotlib

**Comment by eickenberg** [comment]
Simply not calling `set_tick_labels` causes the colorbar ticks to be formatted in a reasonable manner automatically.

**Comment by lesteve** [comment]
@eickenberg you have a tiny merge conflict in a docstring, could you fix it?

**Comment by eickenberg** [comment]
resolved conflict and squashed commits

**Comment by eickenberg** [comment]
Yeah I somehow reverted to an earlier version, all colorbar logic is back in plot_stat_map instead of _plot_img_with_bg.

**Comment by eickenberg** [comment]
OK this doesn't work for the travis machine with `MATPLOTLIB_VERSION="1.4.2"`, a newer matplotlib than that which I used to code this feature. I am having troubles installing matplotlib from github right now, so I'll try to fix this on another machine.

**Comment by eickenberg** [comment]
In matplotlib 1.1.1 we have
`ColorbarBase.outline.get_ydata` and `ColorbarBase.outline.set_ydata`

In matplotlib 1.4.2, that doesn't exist, but there is
`ColorbarBase.outline.get_xy` and `ColorbarBase.outline.set_xy`.

There exists no function to this effect that is shared between versions. One way out of this would be to write helper functions `_get_y, _set_y` that render these differences transparent. Should I do that, or is there something smarter I can do?

**Comment by lesteve** [comment]
> One way out of this would be to write helper functions _get_y, _set_y that render these differences transparent

Sounds good, just put it in nilearn._utils.fixes where this kind of version dependent things tend to live.

**Comment by eickenberg** [comment]
Tests passing on matplotlib 1.4.2 as well now.

File: nilearn/_utils/fixes/matplotlib_backports.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+    Polygon(Patch) object instead of a Line2D(Line) object. This entails
+    different getters and setters.
+
+    Change specifically after commit 48f594c2e2b05839ea394040b06196f39d9fbfba,
+    entitled
+    "changed colorbar outline from a Line2D object to a Polygon object"
+    from August 28th, 2013.
+
+    This function unifies getters and setters of ColorbarBase outline xy
+    coordinates."""
```
Comment (AlexandreAbraham) [inline nilearn/_utils/fixes/matplotlib_backports.py]: Thanks for all these details!

File: nilearn/plotting/img_plotting.py [hunk 3]
```diff
@@ -587,7 +613,15 @@ def plot_stat_map(stat_map_img, bg_img=MNI152TEMPLATE, cut_coords=None,
                     np.logical_not(stat_map_data._mask)])
         stat_map_max = np.nanmax(stat_map_data)
         stat_map_min = np.nanmin(stat_map_data)
+
+    if symmetric_cbar == 'auto':
+        symmetric_cbar = stat_map_min < 0 and stat_map_max > 0
```
Comment (AlexandreAbraham) [inline nilearn/plotting/img_plotting.py]: I prefer to enclose boolean expression in parenthesis `
symmetric_cbar = (stat_map_min < 0 and stat_map_max > 0)`

File: nilearn/plotting/img_plotting.py [hunk 4]
```diff
@@ -595,6 +629,21 @@ def plot_stat_map(stat_map_img, bg_img=MNI152TEMPLATE, cut_coords=None,
                          'the map, use the "threshold" argument')
     vmin = -vmax
 
+    negative_range = stat_map_max <= 0
+    positive_range = stat_map_min >= 0
```
Comment (AlexandreAbraham) [inline nilearn/plotting/img_plotting.py]: Same here.

**Comment by GaelVaroquaux** [comment]
Hey, @eickenberg : sorry we are running out of time, so we are going to release without this. But there will be a new release soon.

Can you confirm: this is an MRG, right?

**Comment by eickenberg** [comment]
Yeah this is/was ready ...

File: nilearn/plotting/img_plotting.py [hunk 5]
```diff
@@ -43,6 +45,7 @@ def _plot_img_with_bg(img, bg_img=None, cut_coords=None,
                       draw_cross=True, black_bg=False,
                       bg_vmin=None, bg_vmax=None, interpolation="nearest",
                       display_factory=get_slicer,
+                      cbar_vmin=None, cbar_vmax=None,
```
Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: Why are why introducing new keywords: could we use 'vmax' and 'threshold'?

Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: @eickenberg convinced.

Comment (GaelVaroquaux) [inline nilearn/plotting/img_plotting.py]: I meant: @eickenberg convinced me.

Comment (eickenberg) [inline nilearn/plotting/img_plotting.py]: ref https://github.com/nilearn/nilearn/pull/293/#issuecomment-68725033

**Comment by GaelVaroquaux** [comment]
Merging. Thanks

**Comment by eickenberg** [comment]
Thanks!

**Comment by GaelVaroquaux** [comment]
> Thanks!

Looks like this merge right before a release was a bit premature :)

**Comment by eickenberg** [comment]
had a little après-sprint to make it grow up :)

**Pull request merged**